File: src/main/scala/com/exasol/cloudetl/scriptclasses/TableDataExporter.scala
package com.exasol.cloudetl.scriptclasses

import scala.collection.mutable.ListBuffer

import com.exasol.ExaIterator
import com.exasol.ExaMetadata
import com.exasol.cloudetl.bucket.Bucket
import com.exasol.cloudetl.data.ExaColumnInfo
import com.exasol.cloudetl.helper.ExasolColumnValueProvider
import com.exasol.cloudetl.sink.BatchSizedSink
import com.exasol.cloudetl.storage.StorageProperties
import com.exasol.common.data.Row

import com.typesafe.scalalogging.LazyLogging

/**
 * A data exporter class that exports table data into a filesystem.
 */
object TableDataExporter extends LazyLogging {

  private[this] val STORAGE_PROPERTIES_INDEX = 1
  private[this] val SOURCE_COLUMNS_INDEX = 2
  private[this] val FIRST_COLUMN_INDEX = 3

  /**
   * Reads the table data and saves them to external filesystem.
   *
   * @param metadata an Exasol metadata object
   * @param iterator an Exasol iterator object
   */
  def run(metadata: ExaMetadata, iterator: ExaIterator): Unit = {
    val storageProperties = StorageProperties(iterator.getString(STORAGE_PROPERTIES_INDEX), metadata)
    val bucket = Bucket(storageProperties)
    val sourceColumnNames = iterator.getString(SOURCE_COLUMNS_INDEX).split("\\.")
    val columns = getColumns(metadata, sourceColumnNames.toSeq)
    val nodeId = metadata.getNodeId()
    val vmId = metadata.getVmId()
    val sink = new BatchSizedSink(nodeId, vmId, iterator.size(), columns, bucket)
    logger.info(s"Starting export from node: $nodeId, vm: $vmId.")
    do {
      sink.write(getRow(iterator, columns))
    } while (iterator.next())
    sink.close()
    iterator.emit(sink.getTotalRecords())
    logger.info(s"Exported '${sink.getTotalRecords()}' records from node '$nodeId' and vm '$vmId'.")
  }

  private[this] def getRow(iterator: ExaIterator, columns: Seq[ExaColumnInfo]): Row = {
    val columnValueProvider = ExasolColumnValueProvider(iterator)
    val values = ListBuffer[Any]()
    var index = 0
    while (index < columns.size) {
      val value = columnValueProvider.getColumnValue(FIRST_COLUMN_INDEX + index, columns(index))
      val _ = values.append(value)
      index += 1
    }
    Row(values.toSeq)
  }

  /**
   * Creates a sequence of [[ExaColumnInfo]] columns using an Exasol
   * [[ExaMetadata]] input column methods.
   *
   * Set the name of the column using `srcColumnNames` parameter.
   * Additionally, set the precision, scale and length using
   * corresponding functions on Exasol metadata for input columns.
   *
   * @param meta An Exasol [[ExaMetadata]] metadata
   * @param sourceColumnNames A sequence of column names per each input column in metadata
   * @return A sequence of [[ExaColumnInfo]] columns
   */
  private[this] def getColumns(meta: ExaMetadata, sourceColumnNames: Seq[String]): Seq[ExaColumnInfo] = {
    val totalColumnCount = meta.getInputColumnCount().toInt
    val columns = ListBuffer[ExaColumnInfo]()
    for { idx <- FIRST_COLUMN_INDEX until totalColumnCount } {
      columns.append(
        ExaColumnInfo(
          name = sourceColumnNames(idx - FIRST_COLUMN_INDEX),
          `type` = meta.getInputColumnType(idx),
          precision = meta.getInputColumnPrecision(idx).toInt,
          scale = meta.getInputColumnScale(idx).toInt,
          length = meta.getInputColumnLength(idx).toInt,
          isNullable = true
        )
      )
    }
    columns.toSeq
  }

}

File: src/main/java/com/exasol/cloudetl/scriptclasses/TableDataExporter.java
package com.exasol.cloudetl.scriptclasses;

import java.util.ArrayList;
import java.util.List;
import java.util.TimeZone;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.exasol.ExaIterator;
import com.exasol.ExaMetadata;
import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.bucket.Bucket;
import com.exasol.cloudetl.data.ExaColumnInfo;
import com.exasol.cloudetl.helper.ExasolColumnValueProvider;
import com.exasol.cloudetl.sink.BatchSizedSink;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.common.data.Row;

/** Exports table data to storage. */
public final class TableDataExporter {
    private static final Logger LOGGER = LoggerFactory.getLogger(TableDataExporter.class);
    private static final int STORAGE_PROPERTIES_INDEX = 1;
    private static final int SOURCE_COLUMNS_INDEX = 2;
    private static final int FIRST_COLUMN_INDEX = 3;

    private TableDataExporter() {
        // Script class.
    }

    /** Run the exporter. */
    public static void run(final ExaMetadata metadata, final ExaIterator iterator)
            throws com.exasol.ExaIterationException, com.exasol.ExaDataTypeException {
        final StorageProperties storageProperties = StorageProperties.apply(iterator.getString(STORAGE_PROPERTIES_INDEX),
                metadata);
        checkIfTimezoneUtc(storageProperties);
        runExport(metadata, iterator, Bucket.create(storageProperties));
    }

    private static void checkIfTimezoneUtc(final StorageProperties storageProperties) {
        if (storageProperties.isEnabled("TIMEZONE_UTC")) {
            TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
        }
    }

    private static void runExport(final ExaMetadata metadata, final ExaIterator iterator, final Bucket bucket)
            throws com.exasol.ExaIterationException, com.exasol.ExaDataTypeException {
        final String[] sourceColumnNames = iterator.getString(SOURCE_COLUMNS_INDEX).split("\\.");
        final scala.collection.immutable.Seq<ExaColumnInfo> columns = getColumns(metadata, sourceColumnNames);
        final long nodeId = metadata.getNodeId();
        final String vmId = metadata.getVmId();
        final BatchSizedSink sink = new BatchSizedSink(nodeId, vmId, iterator.size(), columns, bucket);
        LOGGER.info("Starting export from node: {}, vm: {}.", nodeId, vmId);
        try {
            do {
                sink.write(getRow(iterator, columns));
            } while (iterator.next());
        } finally {
            sink.close();
        }
        final Long totalRecords = sink.getTotalRecords();
        try {
            iterator.emit(totalRecords);
        } catch (final Exception exception) {
            throw new IllegalStateException(exception);
        }
        LOGGER.info("Exported '{}' records from node '{}' and vm '{}'.", totalRecords, nodeId, vmId);
    }

    private static Row getRow(final ExaIterator iterator, final scala.collection.immutable.Seq<ExaColumnInfo> columns) {
        final ExasolColumnValueProvider columnValueProvider = new ExasolColumnValueProvider(iterator);
        final List<Object> values = new ArrayList<>();
        final List<ExaColumnInfo> javaColumns = ScalaConverters.asJavaList(columns);
        for (int index = 0; index < javaColumns.size(); index++) {
            values.add(columnValueProvider.getColumnValue(FIRST_COLUMN_INDEX + index, javaColumns.get(index)));
        }
        return Row.apply(ScalaConverters.seqFromJava(values));
    }

    private static scala.collection.immutable.Seq<ExaColumnInfo> getColumns(final ExaMetadata meta,
            final String[] sourceColumnNames) throws com.exasol.ExaIterationException {
        final int totalColumnCount = (int) meta.getInputColumnCount();
        final List<ExaColumnInfo> columns = new ArrayList<>();
        for (int idx = FIRST_COLUMN_INDEX; idx < totalColumnCount; idx++) {
            columns.add(new ExaColumnInfo(sourceColumnNames[idx - FIRST_COLUMN_INDEX], meta.getInputColumnType(idx),
                    (int) meta.getInputColumnPrecision(idx), (int) meta.getInputColumnScale(idx),
                    (int) meta.getInputColumnLength(idx), true));
        }
        return ScalaConverters.seqFromJava(columns);
    }
}
